### PR TITLE
Add `warnings.deprecated`

### PR DESCRIPTION
**Pull request by LecrisUT**
This is used to add both static type-checking deprecations and run-time as well. For now it requires upgrading `typing-extensions` to a required conditional dependency, otherwise the dynamic deprecation would not apply (and would require some trickery to make it `no-op` on non `TYPE_CHECKING`)

Pull Request Checklist

* [x] implement the feature
* [x] write the documentation
* [x] mention the version
* [x] include a release note

Depends-on #3059, ~~#3117~~ (managed without it), ~~[BGZ#2295924](https://bugzilla.redhat.com/show_bug.cgi?id=2295924)~~

### CODE REVIEW

**Comment by LecrisUT** [comment]
Great, this is a nice issue with pre-commit running a versioned `tmt-lint` instead of the current one from source (dependencies are not being updated)

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -158,6 +158,7 @@ repos:
         # Help installation by reducing the set of inspected botocore release.
         # There is *a lot* of them, and hatch might fetch many of them.
         - "botocore>=1.25.10"      # 1.25.10 is the current one available for RHEL9
+        - "typing-extensions>=4.9.0; python_version < '3.13'"
```
Comment (martinhoyer) [inline .pre-commit-config.yaml:161]: Not related to the PR, but how does this work? Wouldn't it  install all tmt deps anyway?

**Review by martinhoyer — APPROVED** [review]
Do you know if we're able to utilize this for informing users about pending/existing deprecations?

Few places with deprecating stuff:  
https://github.com/teemtee/tmt/blob/0e6863693cee0566271ee3f744d53f334324974b/tmt/utils/__init__.py#L2802  

https://github.com/teemtee/tmt/blob/0e6863693cee0566271ee3f744d53f334324974b/tmt/options.py#L33  

https://github.com/teemtee/tmt/blob/0e6863693cee0566271ee3f744d53f334324974b/tmt/steps/__init__.py#L211  

I'm thinking, deprecated decorator, which would accept 'deprecated in' and 'text' args, which would be visible to end user. Like the --feeling-safe, but easily reusable.

**Comment by LecrisUT** [comment]
> Do you know if we're able to utilize this for informing users about pending/existing deprecations?

Hmm, would need to mull it over. Easily, no, there is none that I could find when navigating through the usage. The implementation seems to rely on native support for deprecation on `click` and `dataclasses` side, so would have to rely on them to figure out how to make a support for type-hinting these.

> I'm thinking, deprecated decorator, which would accept 'deprecated in' and 'text' args, which would be visible to end user. Like the --feeling-safe, but easily reusable.

File: .pre-commit-config.yaml (hunk 1 above)
Comment (LecrisUT) [reply .pre-commit-config.yaml:161]: Yeah, this is a weird situation:
1. pre-commit downloads `tmt 1.35.0` and `pip install` it
2. `python3 -m tmt._pre_commit` is run
3. The code encounters `import tmt`
4. We do not have `src-layout` so instead of `import tmt` going `pre-commit`'s sources, it takes current repo's `./tmt` folder
5. Current source uses `typing-extensions`, but the dependencies were not installed because it was not defined in `tmt 1.35.0`
6. Patch the `.pre-commint-config.yaml` to manually include those dependencies as well

It's a hack that technically works, but it's soo wrong that it does.

File: docs/releases.rst [hunk 2]
```diff
@@ -52,7 +58,6 @@ the ``clean`` subcommand. Users can clean resources from the last
 run or from a run with a given id. Users can also choose to skip
 cleaning ``guests``, ``runs`` or ``images``.
 
-
```
Comment (happz) [inline docs/releases.rst:55]: Seems like an unrelated change.

Comment (LecrisUT) [reply docs/releases.rst:55]: Oops, yeah, rebase issue

**Review by happz — APPROVED** [review]

**Comment by happz** [comment]
Unrelated failures, merging.
